zypp/UserData.h

```diff
@@ -49,6 +49,27 @@ namespace zypp
 
       typedef zypp::ContentType			ContentType;
 
+      ///////////////////////////////////////////////////////////////////
+      /// \brief Erase a key when going out of scope.
+      /// \ingroup g_RAII
+      struct ScopedKey
+      {
+        ScopedKey( UserData & userData_r, std::string key_r )
+        : _userData { userData_r }
+        , _key { std::move(key_r) }
+        {}
+        ScopedKey( UserData & userData_r, std::string key_r, AnyType val_r )
+        : ScopedKey( userData_r, std::move(key_r) )
+        { userData_r.set( _key, std::move(val_r) ); }
+        ~ScopedKey() {
+          _userData.erase( _key );
+        }
+      private:
+        callback::UserData & _userData;
+        std::string _key;
+      };
+      ///////////////////////////////////////////////////////////////////
+
```

Comment:
We have zypp::OnScopeExit for that. Basically a AutoDispose<void> taking a lambda to clean up when leaving a scope. In the upcoming branch i even added some syntactical sugar around it

Reply:
Even better. I'll discard the class.